addon/mirage/index.js

```diff
@@ -43,17 +43,15 @@ export function upload(fn, options={ network: null, timeout: null }) {
           });
 
           metadata.then((metadata) => {
-            let response = {
-              requestBody: Object.assign({
-                [file.key]: metadata
-              }, data)
-            };
+            request.requestBody = Object.assign({
+              [file.key]: metadata
+            }, data);
             if (timedOut) {
               resolve(new Response(408));
               return;
             }
 
-            resolve(fn.call(this, db, response));
+            resolve(fn.call(this, db, request));
```

Comment:
Is it possible to do this without mutating the request object? I prefer a more functional approach with regards to immutability. Helps out with debugging and reasoning of the code :)

Reply:
I'm not sure if I get you right. Are we talking about not setting `responseBody` on `request` (Line 46-48) but constructing a new `request`. I think trying something like that was the root cause for this bug. Instead of passing the instance of `FakeRequest` through, a new pojo was constructed missing a lot of information. For a consumer it's hard to understand why this wrapper function about a route handler alters the route handler's interface. Constructing a new `FakeRequest` instance ourselves would be error prune and I'm not sure if it's even possible as pretender doesn't seem to export that class.

Reply:
We could do something like:
```js
let response = Object.assign({ requestBody: Object.assign({[...]: metadata}, data}, request)
```
Would this work?

Reply:
Not if we don't want to alter the interface. `request` is an instance of `FakeRequest` not a pojo.

Reply:
I think that you can clone prototype objects using [Object.create](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create).

I think this snipppet works, and doesn't modify the original request.

```suggestion
const clonedRequest = Object.create(request);
clonedRequest.requestBody = Object.assign({ [file.key]: metadata }, data);

if (timedOut) {
  resolve(new Response(408));
  return;
}

resolve(fn.call(this, db, clonedRequest));
```

<img width="560" alt="Screen Shot 2019-10-06 at 2 58 50 PM" src="https://user-images.githubusercontent.com/36919/66263171-a7af0580-e84a-11e9-823e-e54fb8d8ff7c.png">

Reply:
Maybe we should not touch the request object at all but provide the file metadata as an additional argument to the route handler? 

```js
resolve(fn.call(this, ...arguments, metadata));
```

Reply:
Hmm. You have convinced me. Lets keep it your way for now. I generally don't like mutating objects if it can be avoided. Maybe here its for the best.